### DISCUSSION TITLE
OpenTopoMap TileLayer?

### DISCUSSION THREAD

**Discussion by LillyWebsites** [Mapsui/Mapsui#1693] [Q&A] [answered]
Hello,
is there a way to use OpenTopoMap as TileLayer?
For OpenStreetMap it works with OpenStreetMap.CreateTileLayer().

For OpenTopoMap I only found these TileServers:
http://a.tile.opentopomap.org/${z}/${x}/${y}.png
http://b.tile.opentopomap.org/${z}/${x}/${y}.png
http://c.tile.opentopomap.org/${z}/${x}/${y}.png

I dont know if this is possible at all. I cannot find samples in Google.
Thanks for any help

**Comment by FrozDark** [comment, marked as answer]
Try
```

    internal class TopoMapTiles : ITileSource, ITileProvider, IRequest, IDisposable
    {
        public ITileSchema Schema { get => new GlobalSphericalMercator(1, 18, "TopoMaps"); }
        public string Name { get => "TopoMaps"; }
        public Attribution Attribution { get; } = new Attribution("© OpenStreetMap contributors", "https://www.openstreetmap.org/copyright");

        private readonly HttpClient _httpClient = HttpClientBuilder.Build();

        private readonly IRequest _request;

        public TopoMapTiles()
        {
            _request = new BasicRequest("https://{s}.tile.opentopomap.org/{z}/{x}/{y}.png", new[] { "a", "b", "c" }, null);
        }

        public byte[] GetTile(TileInfo tileInfo)
        {
            try
            {
                array = _httpClient.GetByteArrayAsync(_request.GetUri(tileInfo)).ConfigureAwait(false).GetAwaiter().GetResult();

                return array;
            }
            catch
            {
                return null;
            }
        }

        public Uri GetUri(TileInfo info)
        {
            return _request.GetUri(info);
        }

        public void Dispose()
        {
            _httpClient?.Dispose();
        }
    }
```

**Reply by LillyWebsites** [reply]
thanks, after fixing constructor name and changing to https it works!

**Reply by LillyWebsites** [reply]
perhaps I was a bit too quick with my answer. It kind of works, but extremely slow. When the activity loads, I see the whole world. When the app automatically zooms in, I only see yellow for around 30 seconds. Setting an user agent as stated in openstreetmap guide did not help
`                _httpClient.DefaultRequestHeaders.Add("User-Agent", Utils.AppName+"/"+Utils.AppVersion.Replace("V ", ""));`
Any idea what might be the reason?

THis shorter code (without derived class) does the same, has the same speed problem:

` var tileSource = new HttpTileSource(new GlobalSphericalMercator(),
                            "https://{s}.tile.opentopomap.org/{z}/{x}/{y}.png",
                            new[] { "a", "b", "c" }, name: "OpenStreetMap",
                            attribution: new BruTile.Attribution("© OpenStreetMap/OpenTopoMap contributors", "http://www.openstreetmap.org/copyright"));
                        map.Layers.Add(new TileLayer(tileSource) { Name = "OpenTopoMap" });`

**Reply by FrozDark** [reply]
I think it's related to the server being slow. Try to use some cache then
